[export] construct set_grad_enabled HOO subgraph inside other HOO subgraphs

pianpwk opened the pull request:
Summary:
Reference: https://github.com/pytorch/pytorch/pull/121736

Previously set_grad_enabled nodes in HOO subgraphs (e.g. cond) were inlined and not replaced with their own HOO subgraphs. This diff recursively does that.

Example:
```
class Model(torch.nn.Module):
    def forward(self, x, y):
        def true_fn(x, y):
            with torch.enable_grad():
                return x - y

        return torch.cond(
            x.sum() > 0,
            true_fn,
            lambda x, y: x + y,
            [x, y],
        )
```

Before (printing out `ep.graph_module.true_graph_0`):
```
        class <lambda>(torch.nn.Module):
            def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                # No stacktrace found for following nodes
                _set_grad_enabled = torch._C._set_grad_enabled(True)
                sub: "i64[]" = torch.ops.aten.sub.Tensor(arg0_1, arg1_1);  arg0_1 = arg1_1 = None
                _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
                return (sub,)
```

After:
```
        class GraphModule(torch.nn.Module):
            def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                # No stacktrace found for following nodes
                submod_3 = self.submod_1
                sub: "i64[]" = torch._higher_order_ops.wrap.wrap_with_set_grad_enabled(True, submod_3, arg0_1, arg1_1);  submod_3 = arg0_1 = arg1_1 = None
                return (sub,)

            class GraphModule(torch.nn.Module):
                def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                    # No stacktrace found for following nodes
                    sub: "i64[]" = torch.ops.aten.sub.Tensor(arg0_1, arg1_1);  arg0_1 = arg1_1 = None
                    return sub
```

Differential Revision: D55770138

tugsbayasgalan commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -132,22 +132,34 @@ def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule):
+    for node in new_gm.graph.nodes:
+        if node.op == "get_attr":
+            subgm = getattr(new_gm, node.target)
+            if not isinstance(subgm, torch.fx.GraphModule):
```
Since this is on the lifted graph, i think it is always going to be a graph module. Maybe we should just assert that.

pianpwk replied:
Hm I will recheck. I think the only potential issue is with custom object attributes

tugsbayasgalan commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -132,22 +132,34 @@ def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule):
+            if not isinstance(subgm, torch.fx.GraphModule):
+                continue
+            new_subgm = replace_set_grad_with_hop_pass(subgm)
+            setattr(new_gm, node.target, new_subgm)
```
Should we delete the old subgm too?

tugsbayasgalan commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -132,22 +132,34 @@ def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule):
+                else node
+            ),
+        )
+    else:
```
NIT: Could you make it a utility function? so that we don't have to do if/else branch. Something like:
```
def get_new_graph(gm):
       if need_replacing: return bar(gm)
       return gm
       
```

pianpwk replied:
Thanks for the review! will do

pianpwk committed:
[export] construct set_grad_enabled HOO subgraph inside other HOO subgraphs (#123391)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/123391

Reference: https://github.com/pytorch/pytorch/pull/121736

Previously set_grad_enabled nodes in HOO subgraphs (e.g. cond) were inlined and not replaced with their own HOO subgraphs. This diff recursively does that.

Example:
```
class Model(torch.nn.Module):
    def forward(self, x, y):
        def true_fn(x, y):
            with torch.enable_grad():
                return x - y

        return torch.cond(
            x.sum() > 0,
            true_fn,
            lambda x, y: x + y,
            [x, y],
        )
```

Before (printing out `ep.graph_module.true_graph_0`):
```
        class <lambda>(torch.nn.Module):
            def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                # No stacktrace found for following nodes
                _set_grad_enabled = torch._C._set_grad_enabled(True)
                sub: "i64[]" = torch.ops.aten.sub.Tensor(arg0_1, arg1_1);  arg0_1 = arg1_1 = None
                _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
                return (sub,)
```

After:
```
        class GraphModule(torch.nn.Module):
            def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                # No stacktrace found for following nodes
                submod_3 = self.submod_1
                sub: "i64[]" = torch._higher_order_ops.wrap.wrap_with_set_grad_enabled(True, submod_3, arg0_1, arg1_1);  submod_3 = arg0_1 = arg1_1 = None
                return (sub,)

            class GraphModule(torch.nn.Module):
                def forward(self, arg0_1: "i64[]", arg1_1: "i64[]"):
                    # No stacktrace found for following nodes
                    sub: "i64[]" = torch.ops.aten.sub.Tensor(arg0_1, arg1_1);  arg0_1 = arg1_1 = None
                    return sub
```

Test Plan: test_export test case

Reviewed By: tugsbayasgalan

Differential Revision: D55770138
